nat-lab/tests/test_pmtu.py

```diff
@@ -0,0 +1,201 @@
+import config
+import pytest
+from contextlib import AsyncExitStack
+from helpers import SetupParameters, setup_environment, setup_connections
+from telio import AdapterType
+from telio_features import TelioFeatures
+from utils import testing, stun
+from utils.connection_util import (
+    ConnectionTag,
+    generate_connection_tracker_config,
+    ConnectionLimits,
+)
+from utils.ping import Ping
+from utils.router import IPStack
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    "setup_params",
+    [
+        pytest.param(
+            SetupParameters(
+                ip_stack=IPStack.IPv4,
+                connection_tag=ConnectionTag.DOCKER_CONE_CLIENT_1,
+            )
+        ),
+        # TODO(msz): Disable IPv6 tests since the docker netowrk uses local link addresses
+        #            causing `connect()` call to fail.
+        # pytest.param(
+        #     SetupParameters(
+        #         ip_stack=IPStack.IPv6,
+        #         connection_tag=ConnectionTag.DOCKER_CONE_CLIENT_1,
+        #     )
+        # ),
+    ],
+)
+async def test_pmtu_black_hole(setup_params: SetupParameters) -> None:
+    async with AsyncExitStack() as exit_stack:
+        env = await exit_stack.enter_async_context(
+            setup_environment(exit_stack, [setup_params])
+        )
+        client, *_ = env.clients
+
+        host = (
+            config.PMTU_PROBE_HOST_IP4
+            if setup_params.ip_stack in [IPStack.IPv4, IPStack.IPv4v6]
+            else config.PMTU_PROBE_HOST_IP6
+        )
+
+        await testing.wait_lengthy(client.probe_pmtu(host, 1300))
+
+        # do it second time bacause the kernel might cache something
+        await testing.wait_lengthy(client.probe_pmtu(host, 1300))
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    "setup_params",
+    [
+        pytest.param(
+            SetupParameters(
+                ip_stack=IPStack.IPv4,
+                connection_tag=ConnectionTag.DOCKER_CONE_CLIENT_1,
+            )
+        ),
+        # TODO(msz): Disable IPv6 tests since the docker netowrk uses local link addresses
+        #            causing `connect()` call to fail.
+        # pytest.param(
+        #     SetupParameters(
+        #         ip_stack=IPStack.IPv6,
+        #         connection_tag=ConnectionTag.DOCKER_CONE_CLIENT_1,
+        #     )
+        # ),
+    ],
+)
+async def test_pmtu_with_nexthop(setup_params: SetupParameters) -> None:
+    async with AsyncExitStack() as exit_stack:
+        env = await exit_stack.enter_async_context(
+            setup_environment(exit_stack, [setup_params])
+        )
+        client, *_ = env.clients
+        connection, *_ = [conn.connection for conn in env.connections]
+
+        host = (
+            config.PMTU_PROBE_HOST_IP4
+            if setup_params.ip_stack in [IPStack.IPv4, IPStack.IPv4v6]
+            else config.PMTU_PROBE_HOST_IP6
+        )
+
+        proc = connection.create_process(["/opt/bin/inject-icmp-host-unreachable"])
```

Comment:
I wonder if we are sure that there is no race against "normal" kernel generated Echo responses and the injected ICMP error message.

Reply:
There shouldn't be since we have iptables rule to drop ICMP over 1400 bytes